### PR TITLE
[Nerve] avoid zk reporter/watcher reaping on network disconnection (cont.)

### PR DESCRIPTION
**Pull request by lap1817**
This PR is to follow the similar idea in the PR on Synapse (https://github.com/airbnb/synapse/pull/250).

The current logic in Nerve is that: if an error is raised from the reporter (e.g. zookeeper reporter) when ping?, report_up, report_down, the watcher thread will stop the reporter (so it tries to remove the znode). Then the Nerve main thread will reap the watcher and relaunch it.

This caused unnecessary Nerve status flipping while in short network disconnection. The ping? function on reporter is most frequently called by the watcher thread. E.g., 
  - when a disconnection happens while the the reporter is ping, it raises an exception
  - the reporter then tries to remove the znode which succeeded as the disconnection was short.
  - after the watcher is relaunched, the znode is added again

Flipping nerve status triggers more Synapse read, which make the network saturation more likely to happen.

### Changes
 - reporter.ping?, report_up, report_down will catch ZK connection exceptions.
 - reporter.report_up, reporter.report_down will return a bool value. 
 -- if everything goes good, they return true
 -- if zk connection exception was captured, return false
 -- other exceptions, raise it
 - service watcher now has a a circuit break. a counter keeps tracking repeated check_report failures (as returned by reporter). if that goes beyond a threshold, the run() thread will end.  

### Test
Rspec added/updated
Tested on mango-test with simulated packet loss

Reviewers: 
@igor47  @jolynch @juchem @ken-experiment

### CODE REVIEW

**Comment by lap1817** [comment]
@juchem , I added the logic to bubble up the connection errors when they happened too many. PTAL

**Comment by lap1817** [comment]
@igor47 , PTAL. Rspec are updated to address your previous comments on using context

File: lib/nerve/service_watcher.rb [hunk 1]
```diff
@@ -131,17 +145,34 @@ def check_and_report
       is_up = check?
       log.debug "nerve: current service status for #{@name} is #{is_up.inspect}"
 
+      report_succeeded = true
```
Comment (igor47) [inline lib/nerve/service_watcher.rb]: this initialization seems superfluous

File: spec/lib/nerve/reporter_zookeeper_spec.rb [hunk 2]
```diff
@@ -29,5 +32,109 @@
...
+
+    it "returns true on ping?" do
+      expect(zk).to receive(:exists?) { "zk_path" }.and_return(true)
+      expect(@reporter.ping?).to be true
+    end
+
+    context "when zk.connected? started to return false" do
+      it 'returns false on ping?' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
```
Comment (igor47) [inline spec/lib/nerve/reporter_zookeeper_spec.rb]: move this expectation out into the context block and out of the example blocks:

```ruby
context "when zk.connected? begins to return false" do
  before { expect(zk).to receive(:connected?).and_return(false)

  # examples go here
end
```

File: spec/lib/nerve/reporter_zookeeper_spec.rb [hunk 3]
```diff
@@ -29,5 +32,109 @@
...
+        expect(zk).not_to receive(:delete)
+        expect(@reporter.report_down).to be false
+      end
+    end
+
+    context "when there is a short disconnection" do
+      it 'returns false on ping?' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:exists?).and_raise(ZK::Exceptions::OperationTimeOut)
```
Comment (igor47) [inline spec/lib/nerve/reporter_zookeeper_spec.rb]: likewise this should live in the context, not in the example blocks

File: spec/lib/nerve/reporter_zookeeper_spec.rb [hunk 4]
```diff
@@ -29,5 +32,109 @@
...
+        expect(zk).to receive(:delete).and_raise(ZK::Exceptions::OperationTimeOut)
+        expect(@reporter.report_down).to be false
+      end
+    end
+
+    context "when there is other ZK errors" do
+      it 'raises zk non-connection error on ping?' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:exists?).and_raise(ZK::Exceptions::SessionExpired)
```
Comment (igor47) [inline spec/lib/nerve/reporter_zookeeper_spec.rb]: here too

File: lib/nerve/service_watcher.rb (hunk 1 above)
Comment (lap1817) [reply lib/nerve/service_watcher.rb]: this is added here to ensure that check_and_report does not return nil when "is_up == @was_up" and reporter.ping? is successful.

Comment (lap1817) [reply lib/nerve/service_watcher.rb]: so it returns true in this case meaning that zk check is successful and there is no status change

File: spec/lib/nerve/reporter_zookeeper_spec.rb (hunk 3 above)
Comment (lap1817) [reply spec/lib/nerve/reporter_zookeeper_spec.rb]: for report_up and report_down, they actually throw exception from a different zk operation, not zk.exists?. so there is no need to have a common before block

File: spec/lib/nerve/reporter_zookeeper_spec.rb (hunk 4 above)
Comment (lap1817) [reply spec/lib/nerve/reporter_zookeeper_spec.rb]: for report_up and report_down, they actually throw exception from a different zk operation, not zk.exists?. so there is no need to have a common before block

File: spec/lib/nerve/reporter_zookeeper_spec.rb (hunk 2 above)
Comment (lap1817) [reply spec/lib/nerve/reporter_zookeeper_spec.rb]: done

**Comment by lap1817** [comment]
@igor47 , addressed the comments. please take a look

**Comment by lap1817** [comment]
@igor47 , PTAL

File: lib/nerve/service_watcher.rb (hunk 1 above)
Comment (igor47) [reply lib/nerve/service_watcher.rb]: to make the intent clear, i would do something like:

```ruby
return true if is_up == @was_up

# if we got here, the state has changed
rest_of_the_code
```

but NBD

**Comment by igor47** [comment]
the change and accompanying specs look much better, thanks

**Pull request merged**